File: framework/C/src/channel.c
/* channel.c -- deterministic synthesis/sequencing error channel */
#include "vivi/channel.h"
#include <string.h>

/* absorbs messages when a caller passes err == nullptr */
static VIVI_THREAD_LOCAL const char *channel_err_sink;

static int digit_at(const uint8_t *s, size_t i)
{
	return (int)((s[i / 4] >> (2 * (3 - i % 4))) & 3u);
}

static bool channel_read_impl(vivi_read *out, const uint8_t *strand, size_t slen,
	const vivi_channel_opts *opts, int soft, const char **err)
{
	memset(out, 0, sizeof(*out));
	if (!strand) { *err = "expected string"; return false; }
	vivi_channel_opts def = { 0.0, 0.0, 0.0, 0.0, 0, 0.0, 0.0, 0.0, 0.0, 0, 0.0 };
	if (!opts) opts = &def;
	const double ps[9] = { opts->p_sub, opts->p_ins, opts->p_del, opts->p_drop,
		opts->p_sub_gc, opts->p_sub_hp, opts->p_trunc, opts->p_burst,
		opts->p_burst_del };
	for (int i = 0; i < 9; i++)
		if (!(ps[i] >= 0.0 && ps[i] <= 1.0)) {
			*err = "invalid probability";
			return false;
		}
	if (slen > (((size_t)-1) - 4) / 8) { *err = "strand too long"; return false; }
	vivi_prng rng;
	vivi_prng_init(&rng, opts->seed);
	if (	vivi_prng_chance(&rng, opts->p_drop)) {
		out->dropped = 1;
		return true;
	}
	size_t n = slen * 4;
	size_t limit = 0;
	if (opts->p_trunc > 0.0 && n >= 2 && vivi_prng_chance(&rng, opts->p_trunc))
		limit = 1 + (size_t)(vivi_prng_next(&rng) % (uint32_t)(n - 1));
	uint8_t *digits = vivi_alloc(2 * n + 4);
	if (!digits) { *err = "out of memory"; return false; }
	uint8_t *q = soft ? vivi_alloc(2 * n + 4) : nullptr;
	if (soft && !q) { vivi_dealloc(digits); *err = "out of memory"; return false; }
	size_t m = 0;
	uint32_t burst_rem = 0;
	for (size_t i = 0; i < n; i++) {
		if (limit && m >= limit) break;
		int d = digit_at(strand, i);
		int bursted = 0;
		if (opts->p_burst > 0.0) {
			if (burst_rem > 0) {
				bursted = 1;
				burst_rem--;
			} else if (	vivi_prng_chance(&rng, opts->p_burst)) {
				burst_rem = opts->burst_len ? opts->burst_len : 8;
				bursted = 1;
			}
		}
		if (!	vivi_prng_chance(&rng, opts->p_del)) {
			double p_sub = opts->p_sub;
			int killed = 0;
			if (bursted) {
				if (opts->p_burst_del > 0.0
					&& 	vivi_prng_chance(&rng, opts->p_burst_del))
					killed = 1;
				else
					p_sub = 1.0;
			} else {
				if (opts->p_sub_gc > 0.0 && (d == 1 || d == 2))
					p_sub = 1.0 - (1.0 - p_sub) * (1.0 - opts->p_sub_gc);
				if (opts->p_sub_hp > 0.0 && i > 0 && d == digit_at(strand, i - 1))
					p_sub = 1.0 - (1.0 - p_sub) * (1.0 - opts->p_sub_hp);
			}
			if (!killed) {
				int sub = 	vivi_prng_chance(&rng, p_sub);
				if (sub) {
					int k = (int)(	vivi_prng_next(&rng) % 3u);
					d = (d + 1 + k) % 4;   /* never the original base */
				}
				if (!limit || m < limit) {
					digits[m] = (uint8_t)d;
					if (q) q[m] = (uint8_t)(sub ? VIVI_QUAL_LO : VIVI_QUAL_HI);
					m++;
				}
			}
		}
		if (	vivi_prng_chance(&rng, opts->p_ins)) {
			uint8_t ins = (uint8_t)(	vivi_prng_next(&rng) & 3u);
			if (!limit || m < limit) {
				digits[m] = ins;
				if (q) q[m] = VIVI_QUAL_LO;
				m++;
			}
		}
	}
	size_t bytes = (m + 3) / 4;
	if (bytes) {
		uint8_t *packed = vivi_zalloc(bytes, 1);
		if (!packed) {
			vivi_dealloc(digits);
			vivi_dealloc(q);
			*err = "out of memory";
			return false;
		}
		for (size_t i = 0; i < m; i++)
			packed[i / 4] |= (uint8_t)(digits[i] << (2 * (3 - i % 4)));
		out->strand.data = packed;
		out->strand.len = bytes;
	}
	out->bases = m;
	vivi_dealloc(digits);
	if (soft) {
		out->qual = vivi_alloc(m ? m : 1);
		if (!out->qual) {
			vivi_bytes_free(&out->strand);
			vivi_dealloc(q);
			*err = "out of memory";
			return false;
		}
		if (m) memcpy(out->qual, q, m);
	}
	vivi_dealloc(q);
	return true;
}

bool vivi_channel_read(vivi_read *out, const uint8_t *strand, size_t slen,
	const vivi_channel_opts *opts, const char **err)
{
	if (!err) err = &channel_err_sink;
	if (!out) { *err = "expected output"; return false; }
	return channel_read_impl(out, strand, slen, opts, 0, err);
}

bool vivi_channel_read_soft(vivi_read *out, const uint8_t *strand, size_t slen,
	const vivi_channel_opts *opts, const char **err)
{
	if (!err) err = &channel_err_sink;
	if (!out) { *err = "expected output"; return false; }
	return channel_read_impl(out, strand, slen, opts, 1, err);
}

void vivi_read_free(vivi_read *out)
{
	if (!out) return;
	vivi_bytes_free(&out->strand);
	vivi_dealloc(out->qual);
	out->qual = nullptr;
	out->bases = 0;
	out->dropped = 0;
}

bool vivi_read_from_bytes(vivi_read *out, const uint8_t *strand, size_t slen,
	size_t bases, const char **err)
{
	if (!err) err = &channel_err_sink;
	if (out) *out = (vivi_read){ 0 };
	if (!out || (!strand && slen)) { *err = "expected buffers"; return false; }
	if (slen < (bases + 3) / 4) { *err = "strand too short for base count"; return false; }
	if (slen) {
		out->strand.data = vivi_alloc(slen);
		if (!out->strand.data) { *err = "out of memory"; return false; }
		memcpy(out->strand.data, strand, slen);
	}
	out->strand.len = slen;
	out->bases = bases;
	return true;
}

uint8_t vivi_read_base(const vivi_read *r, size_t i)
{
	if (!r || !r->strand.data || i >= r->bases) return 0;
	return (uint8_t)((r->strand.data[i / 4] >> (2 * (3 - i % 4))) & 3u);
}

bool vivi_read_alloc_quality(vivi_read *r, uint8_t fill, const char **err)
{
	if (!err) err = &channel_err_sink;
	if (!r) { *err = "expected read"; return false; }
	if (!r->qual) {
		r->qual = vivi_alloc(r->bases ? r->bases : 1);
		if (!r->qual) { *err = "out of memory"; return false; }
	}
	if (r->bases) memset(r->qual, fill, r->bases);
	return true;
}

bool vivi_read_set_quality(vivi_read *r, size_t i, uint8_t qual)
{
	if (!r || !r->qual || i >= r->bases) return false;
	r->qual[i] = qual;
	return true;
}

bool vivi_consensus_vote(vivi_read *out, const vivi_read *const *reads, size_t count,
	int soft, const char **err)
{
	if (!err) err = &channel_err_sink;
	if (!out) { *err = "expected output"; return false; }
	memset(out, 0, sizeof(*out));
	if (!reads && count) { *err = "expected reads"; return false; }
	size_t n = 0;
	int have = 0;
	for (size_t r = 0; r < count; r++) {
		if (!reads[r] || reads[r]->dropped) continue;
		if (!have) { n = reads[r]->bases; have = 1; }
		else if (reads[r]->bases != n) {
			*err = "consensus requires equal-length reads (no indels)";
			return false;
		}
	}
	if (!have) { out->dropped = 1; return true; }
	uint32_t *counts = vivi_zalloc(n ? n : 1, 4 * sizeof(uint32_t));
	if (!counts) { *err = "out of memory"; return false; }
	for (size_t r = 0; r < count; r++) {
		const vivi_read *rd = reads[r];
		if (!rd || rd->dropped) continue;
		if (rd->bases != n || (n && !rd->strand.data)) {
			vivi_dealloc(counts);
			*err = "consensus requires equal-length reads (no indels)";
			return false;
		}
		for (size_t i = 0; i < n; i++) {
			uint32_t w = 1;
			if (soft && rd->qual) w = (uint32_t)rd->qual[i] + 1u;
			counts[i * 4 + (size_t)vivi_read_base(rd, i)] += w;
		}
	}
	size_t bytes = (n + 3) / 4;
	uint8_t *packed = vivi_zalloc(bytes ? bytes : 1, 1);
	if (!packed) { vivi_dealloc(counts); *err = "out of memory"; return false; }
	for (size_t i = 0; i < n; i++) {
		size_t best = 0;
		for (size_t d = 1; d < 4; d++)
			if (counts[i * 4 + d] > counts[i * 4 + best]) best = d;
		packed[i / 4] |= (uint8_t)(best << (2 * (3 - i % 4)));
	}
	out->strand.data = packed;
	out->strand.len = bytes;
	out->bases = n;
	vivi_dealloc(counts);
	return true;
}

bool vivi_consensus_read(vivi_read *out, const uint8_t *strand, size_t slen,
	const vivi_consensus_opts *opts, const char **err)
{
	if (!err) err = &channel_err_sink;
	if (!out) { *err = "expected output"; return false; }
	if (!opts) return vivi_channel_read(out, strand, slen, nullptr, err);
	int soft = opts->soft != 0;
	if (opts->coverage <= 1)
		return soft ? vivi_channel_read_soft(out, strand, slen, &opts->ch, err)
			: vivi_channel_read(out, strand, slen, &opts->ch, err);
	if (!strand) { *err = "expected string"; return false; }
	if (slen > ((size_t)-1) / 4) { *err = "strand too long"; return false; }
	size_t n = slen * 4;
	if (n == 0)
		return soft ? vivi_channel_read_soft(out, strand, slen, &opts->ch, err)
			: vivi_channel_read(out, strand, slen, &opts->ch, err);
	uint32_t *counts = vivi_zalloc(n, 4 * sizeof(uint32_t));
	if (!counts) { *err = "out of memory"; return false; }
	size_t survivors = 0;
	for (uint32_t r = 0; r < opts->coverage; r++) {
		vivi_channel_opts ch = opts->ch;
		ch.seed = opts->ch.seed + r;
		vivi_read rd;
		if (!(soft ? vivi_channel_read_soft(&rd, strand, slen, &ch, err)
			: vivi_channel_read(&rd, strand, slen, &ch, err))) {
			vivi_dealloc(counts);
			return false;
		}
		if (rd.dropped) { vivi_read_free(&rd); continue; }
		if (rd.bases != n) {
			vivi_read_free(&rd);
			vivi_dealloc(counts);
			*err = "consensus requires equal-length reads (no indels)";
			return false;
		}
		for (size_t i = 0; i < n; i++) {
			uint32_t w = 1;
			if (soft && rd.qual) w = (uint32_t)rd.qual[i] + 1u;
			counts[i * 4 + (size_t)digit_at(rd.strand.data, i)] += w;
		}
		vivi_read_free(&rd);
		survivors++;
	}
	memset(out, 0, sizeof(*out));
	if (survivors == 0) {
		vivi_dealloc(counts);
		out->dropped = 1;
		return true;
	}
	size_t bytes = (n + 3) / 4;
	uint8_t *packed = vivi_zalloc(bytes, 1);
	if (!packed) {
		vivi_dealloc(counts);
		*err = "out of memory";
		return false;
	}
	for (size_t i = 0; i < n; i++) {
		size_t best = 0;
		for (size_t d = 1; d < 4; d++)
			if (counts[i * 4 + d] > counts[i * 4 + best]) best = d;
		packed[i / 4] |= (uint8_t)(best << (2 * (3 - i % 4)));
	}
	out->strand.data = packed;
	out->strand.len = bytes;
	out->bases = n;
	vivi_dealloc(counts);
	return true;
}

File: framework/luau/channel.lua
--!native
local band, bxor, rshift, lshift, bor =
	bit32.band, bit32.bxor, bit32.rshift, bit32.lshift, bit32.bor
local floor = math.floor
local byte, char, concat = string.byte, string.char, table.concat

local TWO32 = 4294967296
local TWO16 = 65536
local ADD_HI, ADD_LO = 0x9E3779B9, 0x7F4A7C15
local M1_HI, M1_LO = 0xBF58476D, 0x1CE4E5B9
local M2_HI, M2_LO = 0x94D049BB, 0x133111EB

local function add64(a, b)
	local lo = a.lo + b.lo
	local hi = a.hi + b.hi
	if lo >= TWO32 then
		lo = lo - TWO32
		hi = hi + 1
	end
	return { hi = hi % TWO32, lo = lo }
end

local function xor64(a, b)
	return { hi = bxor(a.hi, b.hi), lo = bxor(a.lo, b.lo) }
end

local function shr64(a, n)
	if n == 0 then
		return { hi = a.hi, lo = a.lo }
	elseif n < 32 then
		return { hi = rshift(a.hi, n), lo = bor(rshift(a.lo, n), lshift(a.hi, 32 - n)) }
	elseif n == 32 then
		return { hi = 0, lo = a.hi }
	end
	return { hi = 0, lo = rshift(a.hi, n - 32) }
end

local function mul64(a, mhi, mlo)
	local a0, a1 = band(a.lo, 0xFFFF), rshift(a.lo, 16)
	local a2, a3 = band(a.hi, 0xFFFF), rshift(a.hi, 16)
	local b0, b1 = band(mlo, 0xFFFF), rshift(mlo, 16)
	local b2, b3 = band(mhi, 0xFFFF), rshift(mhi, 16)
	local p0 = a0 * b0
	local p1 = a0 * b1 + a1 * b0
	local p2 = a0 * b2 + a1 * b1 + a2 * b0
	local p3 = a0 * b3 + a1 * b2 + a2 * b1 + a3 * b0
	local r0 = p0 % TWO16
	local c = floor(p0 / TWO16)
	local t1 = p1 + c
	local r1 = t1 % TWO16
	c = floor(t1 / TWO16)
	local t2 = p2 + c
	local r2 = t2 % TWO16
	c = floor(t2 / TWO16)
	local t3 = p3 + c
	local r3 = t3 % TWO16
	return { hi = r2 + r3 * TWO16, lo = r0 + r1 * TWO16 }
end

local function prng_new(seed)
	return { state = { hi = 0, lo = seed % TWO32 } }
end

local function prng_mix(p)
	p.state = add64(p.state, { hi = ADD_HI, lo = ADD_LO })
	local z = p.state
	z = xor64(z, shr64(z, 30))
	z = mul64(z, M1_HI, M1_LO)
	z = xor64(z, shr64(z, 27))
	z = mul64(z, M2_HI, M2_LO)
	return xor64(z, shr64(z, 31))
end

local function prng_next(p)
	return prng_mix(p).hi
end

local function prng_next64(p)
	return prng_mix(p)
end

local function prng_chance(p, prob)
	if prob <= 0.0 then return false end
	if prob >= 1.0 then return true end
	local thr = floor(prob * 18446744073709551616.0)
	local r = prng_mix(p)
	local thi = floor(thr / TWO32)
	if r.hi ~= thi then return r.hi < thi end
	return r.lo < thr % TWO32
end

local function digit_at(s, i)
	return band(rshift(byte(s, floor(i / 4) + 1), 2 * (3 - i % 4)), 3)
end

local function pack_digits(digits, m)
	if m == 0 then return "" end
	local bytes = floor((m + 3) / 4)
	local out = {}
	for i = 1, bytes do out[i] = 0 end
	for i = 0, m - 1 do
		local k = floor(i / 4) + 1
		out[k] = bor(out[k], lshift(digits[i + 1], 2 * (3 - i % 4)))
	end
	local t = {}
	for i = 1, bytes do t[i] = char(out[i]) end
	return concat(t)
end

local function channel_read(strand, opts)
	if type(strand) ~= "string" then return nil, "expected string" end
	opts = opts or {}
	local p_sub = opts.p_sub or 0.0
	local p_ins = opts.p_ins or 0.0
	local p_del = opts.p_del or 0.0
	local p_drop = opts.p_drop or 0.0
	local seed = opts.seed or 0
	local p_sub_gc = opts.p_sub_gc or 0.0
	local p_sub_hp = opts.p_sub_hp or 0.0
	local p_trunc = opts.p_trunc or 0.0
	local p_burst = opts.p_burst or 0.0
	local burst_len = opts.burst_len or 0
	local ps = { p_sub, p_ins, p_del, p_drop, p_sub_gc, p_sub_hp, p_trunc, p_burst }
	for i = 1, 8 do
		if not (ps[i] >= 0.0 and ps[i] <= 1.0) then
			return nil, "invalid probability"
		end
	end
	local rng = prng_new(seed)
	if prng_chance(rng, p_drop) then
		return { strand = "", bases = 0, dropped = 1 }
	end
	local n = #strand * 4
	local limit = 0
	if p_trunc > 0.0 and n >= 2 and prng_chance(rng, p_trunc) then
		limit = 1 + prng_next(rng) % (n - 1)
	end
	local digits = {}
	local m = 0
	local burst_rem = 0
	for i = 0, n - 1 do
		if limit ~= 0 and m >= limit then break end
		local d = digit_at(strand, i)
		local bursted = false
		if p_burst > 0.0 then
			if burst_rem > 0 then
				bursted = true
				burst_rem = burst_rem - 1
			elseif prng_chance(rng, p_burst) then
				burst_rem = (burst_len ~= 0) and burst_len or 8
				bursted = true
			end
		end
		if not prng_chance(rng, p_del) then
			local psub = p_sub
			if bursted then
				psub = 1.0
			else
				if p_sub_gc > 0.0 and (d == 1 or d == 2) then
					psub = 1.0 - (1.0 - psub) * (1.0 - p_sub_gc)
				end
				if p_sub_hp > 0.0 and i > 0 and d == digit_at(strand, i - 1) then
					psub = 1.0 - (1.0 - psub) * (1.0 - p_sub_hp)
				end
			end
			if prng_chance(rng, psub) then
				local k = prng_next(rng) % 3
				d = (d + 1 + k) % 4
			end
			if limit == 0 or m < limit then
				m = m + 1
				digits[m] = d
			end
		end
		if prng_chance(rng, p_ins) then
			local ins = prng_next(rng) % 4
			if limit == 0 or m < limit then
				m = m + 1
				digits[m] = ins
			end
		end
	end
	return { strand = pack_digits(digits, m), bases = m, dropped = 0 }
end

local function consensus_read(strand, opts)
	if opts == nil then return channel_read(strand, nil) end
	local ch = opts.ch
	local coverage = opts.coverage or 0
	if coverage <= 1 then return channel_read(strand, ch) end
	if type(strand) ~= "string" then return nil, "expected string" end
	local n = #strand * 4
	if n == 0 then return channel_read(strand, ch) end
	local counts = {}
	for i = 0, n * 4 - 1 do counts[i] = 0 end
	local survivors = 0
	for r = 0, coverage - 1 do
		local c = {}
		if ch then for k, v in pairs(ch) do c[k] = v end end
		c.seed = ((ch and ch.seed or 0) + r) % TWO32
		local rd, err = channel_read(strand, c)
		if not rd then return nil, err end
		if rd.dropped ~= 1 then
			if rd.bases ~= n then
				return nil, "consensus requires equal-length reads (no indels)"
			end
			for i = 0, n - 1 do
				local idx = i * 4 + digit_at(rd.strand, i)
				counts[idx] = counts[idx] + 1
			end
			survivors = survivors + 1
		end
	end
	if survivors == 0 then
		return { strand = "", bases = 0, dropped = 1 }
	end
	local best = {}
	for i = 0, n - 1 do
		local b = 0
		for d = 1, 3 do
			if counts[i * 4 + d] > counts[i * 4 + b] then b = d end
		end
		best[i + 1] = b
	end
	return { strand = pack_digits(best, n), bases = n, dropped = 0 }
end

return {
	read = channel_read,
	consensus_read = consensus_read,
	prng_new = prng_new,
	prng_next = prng_next,
	prng_next64 = prng_next64,
	prng_chance = prng_chance,
}
